src/WorldIDIdentityManagerImplV1.sol

```diff
@@ -782,17 +782,102 @@ contract WorldIDIdentityManagerImplV1 is WorldIDImpl, IWorldID {
         view
         virtual
     {
-        bool previousIsZero = false;
-
-        for (uint256 i = 0; i < identityCommitments.length; ++i) {
-            uint256 commitment = identityCommitments[i];
-            if (previousIsZero && commitment != EMPTY_LEAF) {
-                revert InvalidCommitment(i);
+        // this will be assigned an unreduced element if one is found
+        // note that zero is never unreduced, so it's a safe no-revert state
+        uint256 revertOnUnreduced = 0;
+        // this will be assigned a non-zero element index that occurs after a zero element
+        // note that this definition means an actual revert always happens at index at least 1,
+        // so zero is a safe no-revert state
+        uint256 revertOnInvalid = 0;
+        assembly {
+            // offset one byte after the end of argument section
+            let max := add(mul(identityCommitments.length, 32), identityCommitments.offset)
+            // offset of the first element in the array
+            let offset := identityCommitments.offset
+
+            // increment offset until either of the following happens:
+            // - offset is equal to max, meaning we've reached the end of the array
+            // - the element at offset is zero
+            // - the element at offset is greater than or equal to SNARK_SCALAR_FIELD
+            for
+                { }
+                and(lt(offset, max), and(lt(calldataload(offset), SNARK_SCALAR_FIELD), sub(iszero(calldataload(offset)), 1)))
```

Comment:
`sub(iszero(calldataload(offset)), 1)` 

It's more common and cheaper to do `iszero(iszero(calldataload(offset)))`.